Add example for Widget window in Qt Quick

Cuperino opened the pull request:
Example app where a window bounces across the screen, like an old DVD player's logo would.  You control the text through one of two identical forms, one implemented in QML and the other in Widgets.  The example code follows best practices.

Please note that because the app doesn't work on Wayland because it manipulates the window's position. If you try it on Linux, make sure to run X11.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/widgetFormHandler.cpp:
```diff
@@ -0,0 +1,42 @@
+    emit visibleChanged();
+}
+
+QString WidgetFormHandler::getText()
```
Getters should be const methods (this applies to all 3 getters -- and to other classes in this MR too)

Qt's best practice is to call such a getter `text()` rather than `getText()`, maybe we want to apply the same here?

Cuperino replied:
Done

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontsbackend.cpp:
```diff
@@ -0,0 +1,13 @@
+{
+}
+
+QStringList FontsBackend::fontList()
```
const method (or even static, if that works for QML too)

Cuperino replied:
Done. Set static const.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.h:
```diff
@@ -0,0 +1,40 @@
+    void closeEvent(QCloseEvent *event);
+
+private slots:
+    void on_lineEdit_textChanged(const QString &arg1);
```
`on_*` is definitely not best practice (no error at compile time in case of a mismatch). Please use a real connect to PMF.

Cuperino replied:
Switched to explicit, modern, connects. Removed confusing `on_` prefix.
Are my declarations and casing correct or can these be improved?

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.h:
```diff
@@ -0,0 +1,40 @@
+    void on_pushButton_clicked();
+
+protected:
+    void closeEvent(QCloseEvent *event);
```
override

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/widgetFormHandler.cpp:
```diff
@@ -0,0 +1,42 @@
+    if (visible)
+        m_window->show();
+    else
+        m_window->hide();
```
m_window->setVisible(visible);

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.cpp:
```diff
@@ -0,0 +1,47 @@
+    emit textChanged();
+}
+
+const QString FontControlsWidgetsForm::font()
```
Misunderstanding.
Returning a const QString serves no purpose (on the contrary - it prevents move semantics for the caller)

When I said "make those methods const", I meant marking the actual methods themselves as const:

QString FontControlsWidgetsForm::font() const

(both in .h and .cpp)

(applies to *all* getters)

The purpose is to indicate (and ensure) that the member variables are not (and cannot be) modified in such methods, because in a const method, all member variables are const.

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.cpp:
```diff
@@ -0,0 +1,47 @@
+    return ui->fontComboBox->currentFont().family();
+}
+
+void FontControlsWidgetsForm::closeEvent(QCloseEvent *event) {
```
Inconsistent formatting: you inserted a newline before '{' in most other methods (which is indeed more common).

You *could* grab the _clang-format from e.g. KDABViewer and run clang-format over the whole code, for consistent formatting.

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.cpp:
```diff
@@ -0,0 +1,47 @@
+    QApplication::quit();
+}
+
+void FontControlsWidgetsForm::fontComboBox_currentFontChanged(const QFont &f)
```
OK I see why you asked about casing... This snake/camel mix is weird ;)

Suggestions:
`slotFontChanged`
`onFontChanged`

or in you really need to mention the originating widget, say if you have two lineEdits and you were connecting to `textChanged()` from each, you'd say something like `slotNameChanged` and `slotCountryChanged` or possibly `slotNameTextChanged`/`slotCountryTextChanged`. But it's not like the exact name of the signal has to be in the slot name. Only if you were connecting to multiple similar signals.

Cuperino replied:
Got it. Switched solely to snake case, for consistency.

Cuperino replied:
Removed originating widget name as well. There was no need for distinction.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.h:
```diff
@@ -0,0 +1,40 @@
+signals:
+    void textChanged();
+    void fontChanged();
+    void pushButton_clicked();
```
buttonClicked()

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontsbackend.h:
```diff
@@ -0,0 +1,21 @@
+{
+    Q_OBJECT
+    QML_ELEMENT
+private:
```
remove

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontsbackend.h:
```diff
@@ -0,0 +1,21 @@
+
+    Q_INVOKABLE static const QStringList fontList();
+
+signals:
```
remove

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/timer.h:
```diff
@@ -0,0 +1,23 @@
+public:
+    explicit Timer(QObject *parent = nullptr);
+
+    Q_INVOKABLE int deltaTime();
```
Q_INVOKABLE int deltaTime() const;

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/widgetFormHandler.cpp:
```diff
@@ -0,0 +1,39 @@
+    m_window = new FontControlsWidgetsForm();
+    m_window->setVisible(false);
+
+    QObject::connect(m_window, &FontControlsWidgetsForm::pushButton_clicked, this, &WidgetFormHandler::toggleWidgetsWindow);
```
Remove "QObject::", unneeded, this is a QObject subclass, it's in scope.

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/widgetFormHandler.h:
```diff
@@ -0,0 +1,37 @@
+#ifndef WIDGETFORMHANDLER_H
+#define WIDGETFORMHANDLER_H
+
+#include "fontcontrolswidgetsform.h"
```
isn't `class FontControlsWidgetsForm;` enough in the header?
(and of course, do the real `#include` in the .cpp file)

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/widgetFormHandler.h:
```diff
@@ -0,0 +1,37 @@
+
+#include <QObject>
+#include <QQmlEngine>
+#include <QWidget>
```
not needed here

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/timer.h:
```diff
@@ -0,0 +1,23 @@
+
+    Q_INVOKABLE int deltaTime();
+
+signals:
```
unused, remove

Cuperino replied:
Done.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+
+project(WidgetWindowsInQtQuickApp VERSION 0.1 LANGUAGES CXX)
+
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
This line is usually just after set(CMAKE_CXX_STANDARD 17) or some other value.
Here you're saying "the value of CMAKE_CXX_STANDARD is required" but without saying what CMAKE_CXX_STANDARD is. Either do specify it, or let Qt specify it (Qt 6 requires C++17) in which case you don't need this line, I would think?

Cuperino replied:
Removed. This should make things easier to maintain.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.cpp:
```diff
@@ -0,0 +1,48 @@
+
+FontControlsWidgetsForm::~FontControlsWidgetsForm()
+{
+    delete ui;
```
The problem with saying "This code follows best practices" in the commit log, is that I'll keep spotting anything that doesn't actually follow best practices ;)

The best practice here is to turn `ui` into a std::unique_ptr, and turn this destructor into
`FontControlsWidgetsForm::~FontControlsWidgetsForm() = default;`
(stil here in the .cpp file - don't move it to the header because there the Ui class isn't fully defined and some compilers won't like it)

Cuperino replied:
I'm getting a good refresher in the process. Changed to a unique pointer.

As for the destructor, I noticed that if I use a unique pointer and don't set the default destructor or don't set it the way you indicated, with GCC I will get something like `error: invalid application of ‘sizeof’ to incomplete type ‘FontControlsWidgetsForm’` is that what you were referring to?

dfaure-kdab replied:
Yes, exactly. If the destructor isn't there or is defaulted in the header file, when gcc generates it (in the context of the header file), it doesn't know the size of the type so it doesn't know how much memory to deallocate.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.cpp:
```diff
@@ -0,0 +1,48 @@
+    delete ui;
+}
+
+const QString FontControlsWidgetsForm::text()
```
Returning const isn't best practice, it's even considered harmful because it prevents moves, IIRC.

Cuperino replied:
Fixed.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.cpp:
```diff
@@ -0,0 +1,48 @@
+    return ui->lineEdit->text();
+}
+
+void FontControlsWidgetsForm::lineEditTextChanged(const QString &arg1)
```
s/arg1/text/

Cuperino replied:
Fixed.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.cpp:
```diff
@@ -0,0 +1,48 @@
+    , ui(new Ui::FontControlsWidgetsForm)
+{
+    ui->setupUi(this);
+    connect(ui->lineEdit, &QLineEdit::textEdited, this, &FontControlsWidgetsForm::lineEditTextChanged);
```
Rename the slot to `lineEditTextEdited`, otherwise when I read the slot I thought it was connected to `textChanged` --- but see below, maybe connecting to textChanged would actually be better.

Cuperino replied:
Went with the second option and connected straight to the emitter.
Did this for `currentFontChanged` as well.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.cpp:
```diff
@@ -0,0 +1,48 @@
+void FontControlsWidgetsForm::setText(const QString &text)
+{
+    ui->lineEdit->setText(text);
+    emit textChanged();
```
Actually, if you were connecting to textChanged instead of textEdited, you could remove this emit, it would be done automatically by setText -> textChanged -> lineEditTextChanged()  (obviously, don't rename the slot if you change the signal you connect to)

Cuperino replied:
Nice catch!

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.cpp:
```diff
@@ -0,0 +1,48 @@
+}
+
+void FontControlsWidgetsForm::closeEvent(QCloseEvent *event)
+{
```
Q_UNUSED(event); to avoid a compiler warning

Cuperino replied:
Connected straight to the emitter, eliminating this function.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.cpp:
```diff
@@ -0,0 +1,48 @@
+    QApplication::quit();
+}
+
+void FontControlsWidgetsForm::currentFontChanged(const QFont &f)
```
Q_UNUSED(f); or remove the name `f` from the parameter, which doesn't add anything.

Cuperino replied:
Connected straight to the emitter, eliminating this function.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.h:
```diff
@@ -0,0 +1,40 @@
+    explicit FontControlsWidgetsForm(QWidget *parent = nullptr);
+    ~FontControlsWidgetsForm();
+
+    const QString text();
```
`QString text() const;`
(the method is const, not the QString)

Cuperino replied:
Fixed.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontcontrolswidgetsform.h:
```diff
@@ -0,0 +1,40 @@
+
+    const QString text();
+    void setText(const QString&);
+    QString font() const;
```
fontFamily(), rather?

font() sounds like it will return a QFont....

Cuperino replied:
Switched to using fontFamily where applicable.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/widgetFormHandler.cpp:
```diff
@@ -0,0 +1,41 @@
+    emit visibleChanged();
+}
+
+const QString WidgetFormHandler::text()
```
the method is const, not the QString => move the const to the end

Cuperino replied:
Fixed.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/widgetFormHandler.cpp:
```diff
@@ -0,0 +1,41 @@
+    connect(m_window, &FontControlsWidgetsForm::fontChanged, this, &WidgetFormHandler::fontChanged);
+}
+
+const bool WidgetFormHandler::isVisible()
```
const at the end

Cuperino replied:
Fixed.

dfaure-kdab replied:
nope ;)

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/widgetFormHandler.cpp:
```diff
@@ -0,0 +1,41 @@
+    emit textChanged();
+}
+
+const QString WidgetFormHandler::font()
```
fontFamily()?

(and const at the end, in any case)

Cuperino replied:
Fixed.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/widgetFormHandler.cpp:
```diff
@@ -0,0 +1,41 @@
+WidgetFormHandler::WidgetFormHandler(QObject *parent)
+    : QObject{parent}
+{
+    m_window = new FontControlsWidgetsForm();
```
Who ever deletes this?

Cuperino replied:
Starting today, another unique_ptr.

dfaure-kdab commented on Blog-projects/Widget-window-in-Qt-Quick-app/fontsbackend.cpp:
```diff
@@ -0,0 +1,13 @@
+{
+}
+
+const QStringList FontsBackend::fontFamilyList()
```
const at the end

Cuperino replied:
Static member function cannot have const qualifier. See fontsbackend.h#14

Cuperino replied:
Removed const instead.

dfaure-kdab replied:
Ah, right, yes.

dfaure-kdab reviewed (changes requested):
Almost there... for the C++ code. I can't review the QML code, you'll want to find someone else for that.

Cuperino committed:
Add example for Widget window in Qt Quick

Example app wher a window bounces across the screen,
like an old DVD player's logo would. You control
the text through one of two identical forms,
one implemented in QML and the other in Widgets.
The example code follows best practices.

Due to limitations with Wayland, it only works on
X11 on Linux.

renatofilho commented on Blog-projects/Widget-window-in-Qt-Quick-app/FontControlsQmlForm.qml:
```diff
@@ -0,0 +1,51 @@
+            Layout.fillWidth: true
+        }
+        Button {
+            text: "Switch to Widgets Form"
```
I noticed there is a mix of qStr usage, is that intentional?

Cuperino replied:
Good catch. Nope, I'll fix it.

Cuperino committed:
Use qsTr consistently

Cuperino committed:
Use KDAB brand font

Cuperino committed:
Fix inconsistent call to parent

Cuperino committed:
Add space between QML sections

Improves legibility

Cuperino commented:
I had people from my team, Initech, review the code. All of their suggestions have now been incorporated.
